README.md
# Incremental Data Loader

## Overview
At a high level, this program represents a data integration pipeline. It takes data from flat files and incrementally integrates it into a production table on an SQL Server database. 

<img src="https://github.com/richardogoma/data-integration-pipeline/assets/108296666/77d56164-2c96-43bd-87cd-9ff7f1de1728" alt="Digraph: Data Integration Pipeline Data and Control Flow" style="display: block; margin-left: auto; margin-right: auto;" width="500" height="500">

Here is a high-level summary of what the program does: 

1. **Setup**: The program performs initial setup tasks such as building and running the Docker services, creating configuration files and verifying the existence of the raw data landing zone.

2. **Data Staging**: The program prepares the data for integration by converting its format from CSV to TSV, creating a dynamic temporary table schema, and bulk copying the raw data into the temporary table.

3. **Data Integration**: The program integrates the staged data into the production table. It checks if the production table exists, and if it does, it proceeds to transform the staged data and incrementally load it into the production table, that is, loading only new and updated records only. 

4. **Logging**: The program logs the output of the data integration process and records the program's runtime. It writes the data integration output to a log and also logs the program's runtime.

5. **Error Handling**: The program includes error handling mechanisms. If any errors occur during the execution, they are caught, and the error message is printed to the console.

In summary, this program automates the process of integrating data from flat files into a production table on an SQL Server database, handling data staging, transformation, and logging, while also providing error handling capabilities.

## Use Case
The program can be used to extend the functionality of a program developed to programmatically extract data from an API endpoint as flat files, and to integrate that data into a central repository to support the data management, analytics and BI systems.

## Running the program
* The raw data landing zone is `data\raw` directory
* The Docker service contract is defined in `docker-compose.yml`

    ```bash
    # Setup and run the data integration service
    chmod +x setup.sh && ./setup.sh
    ```
        
> If the program is invoked in another program, considering the use case, the client program should be architected to load data to the raw data landing zone, and the environment variables in `setup.sh` file can be programmatically updated using placeholders.

## Constraints
* The source data could either be in `.csv` or `.tsv` file format. These are commonly used file formats for data transport, and they're non-binary files.
* The primary key has to be _the first field in the source data._ This is a best practice.

    > The primary key should be the first field in your table design. While most databases allow you to define the primary key on any field in your table, the common convention and what the next developer will expect, is the primary key field coming first.

* The primary key field should _be of numeric datatype._ While it is true that there could be some irregularity in the source data, like there could be some strings in the primary key column, it is best practise to use a numeric field as primary key. Preliminary clean-up has been built into the program to expunge rows that are non-numeric in the primary key field; the primary key field is `id`. 

    > Uniquely tagging a record can be done with a number (long integer). Text fields require more bytes than numeric fields, so using a number saves considerable space. Please refer to this article on [Primary Key Tips and Techniques](https://www.fmsinc.com/free/newtips/PrimaryKey.asp)

* The production tables have to be created before executing the program with the correct datatypes for each field. Refer to `.db/init/init.sql`

* The source data must have an `updated_at` field. This is particularly useful for incrementally loading the data to the production table.





docs/history.txt
    1  git status
    2  git fetch origin
    3  git status
    4  git pull origin main
    5  cls
    6  clear
    7  docker-compose up sql-server-db -d
    8  clear
    9  docker-compose up -d sql-server-db
   10  clear
   11  docker-compose ps
   12  docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' dc_sql-server-db
   13  clear
   14  docker-compose up --build --remove-orphans -d
   15  clear
   16  docker-compose up --build --remove-orphans -d
   17  clear
   18  docker exec -it dc_sql-server-db /opt/mssql-tools/bin/sqlcmd -S 172.18.0.2,1433 -U sa -P P@s5wOrd
   19  clear
   20  docker-compose ps -a
   21  docker-compose ls -a
   22  docker-compose down
   23  docker-compose ps -a
   24  docker container ls -a
   25  docker rmi ls -a
   26  docker rmi --help
   27  docker image ls -a
   28  docker image prune
   29  docker image ls -a
   30  clear
   31  docker-compose up --build --remove-orphans -d
   32  clear
   33  docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' dc_sql-server-db
   34  docker-compose ps -a
   35  docker-compose up --build --remove-orphans -d
   36  clear
   37  docker-compose up --build --remove-orphans -d
   38  CLS
   39  cls
   40  clear
   41  echo $USER
   42  clear
   43  docker-compose up --build --remove-orphans -d
   44  clear
   45  docker-compose ps -a
   46  clear
   47  docker-compose down
   48  docker container ls -a
   49  docker image ls -a
   50  docker image prune
   51  clear
   52  docker image ls -a
   53  clear
   54  docker-compose up --build --remove-orphans -d
   55  echo $DB_USER
   56  clear
   57  docker-compose down
   58  docker container ls -a
   59  docker image ls -a
   60  docker image prune
   61  clear
   62  docker-compose up -d sql-server-db
   63  docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' dc_sql-server-db
   64  docker-compose up --build data-integrator --remove-orphans -d
   65  docker-compose up data-integrator --build --remove-orphans -d
   66  docker-compose up --build --remove-orphans -d
   67  echo $User
   68  echo $DB_USER
   69  clear
   70  docker-compose up --build --remove-orphans -d
   71  clear
   72  history
   73  clear
   74  docker container ls -a
   75  docker inspect -f '{{range $p, $conf := .NetworkSettings.Ports}}{{(index $conf 0).HostPort}} -> {{$p}}{{end}}' dc_sql-server-db
   76  clear
   77  docker-compose down
   78  docker container ps -a
   79  docker image ps -a
   80  docker image ls -a
   81  docker image prune
   82  clear
   83  history > history.txt


run.sh
#!/bin/bash

# Build and run the sql-server-db service
docker-compose up -d --remove-orphans sql-server-db

sleep 30

# Print to console sql-server-db-IP-ADDRESS
ip_address=$(docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' dc_sql-server-db)

echo "sql-server-db container ip: $ip_address"

export SQL_SERVER_IP=$ip_address

# Print to console sql-server-db-PORT
# port=$(docker inspect -f '{{range $p, $conf := .NetworkSettings.Ports}}{{(index $conf 0).HostPort}} -> {{$p}}{{end}}' dc_sql-server-db)
port=$(docker inspect -f '{{range $p, $conf := .NetworkSettings.Ports}}{{(index $conf 0).HostPort}}{{end}}' dc_sql-server-db)

echo "sql-server-db container host port: $port"

# Build and run the data-integrator service
docker-compose up -d --remove-orphans data-integrator

sleep 30

# Set the environment variables defined in the .env file
eval $(cat .env | grep SA_PASSWORD)
eval $(cat .env | grep DB_USER)

# Invoke sqlcmd in container db
echo "Run sql commands in terminal ...."
docker exec -it dc_sql-server-db /opt/mssql-tools/bin/sqlcmd -S $ip_address,$port -U $DB_USER -P $SA_PASSWORD

timeout 30 
# --------------------------------
# Run these to confirm data loaded to database table
# select name from sys.databases;
# go

# use testdb;

# select name from sys.tables;
# go

# select top 5 first_name from mock_data;
# go

# exit


src/DataIntegration.ps1
#Requires -Modules @{ ModuleName="SqlServer"; RequiredVersion="21.1.18256" }
#Requires -PSEdition Core
<#
    This script stages and integrates data from a flat file into a SQL Server table based 
    on specified parameters, and logs the results in program and SQL DML log files.
#>

try {
    $rawfilepath = "data/raw/{0}" -f $($env:DATASET)
    $convertedfilepath = "data/processed/{0}" -f $($env:DATASET)
    $InstanceName = $env:SQL_SERVER_IP
    $Database = $env:DATABASE
    $Schema = $env:SCHEMA
    $TableName = $env:TABLE

    # SQL server credential object
    $User = $env:DB_USER
    $PWord = ConvertTo-SecureString -String $($env:SA_PASSWORD) -AsPlainText -Force
    $SqlCred = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $User, $PWord

    # Invoke the data staging function to load data to staging table
    . .\src\utils\fxStageData.ps1
    $staging_response = StageData -InstanceName $InstanceName -Database $Database -Schema $Schema -rawfilepath $rawfilepath `
        -convertedfilepath $convertedfilepath -TableName $TableName -SqlCred $SqlCred # -Verbose
    
    # Invoke the data integration function to integrate new and changed data to the production table
    . .\src\utils\fxIntegrateStagedData.ps1
    $integration_response = IntegrateData -InstanceName $InstanceName -Database $Database -Schema $Schema `
        -TableName $TableName -SqlCred $SqlCred -DataColumns $staging_response['processed_data_headers']
    
    # Write to data integration output to the DML log file (updates & inserts)
    . .\src\utils\fxWriteLog.ps1
    WriteFunction -data $integration_response['log'] -header $integration_response['log_header'] -file ".\log\dml.log"

    # Write to Program log
    $programLogMessage = "Info: Program ran successfully"
    $programLogEntry = [PSCustomObject]@{
        "Timestamp" = Get-Date
        "Instance" = $InstanceName
        "Table" = "[$Database].[$Schema].[$TableName]"
        "LogMessage" = $programLogMessage
        "ScriptName" = $MyInvocation.MyCommand.Name
        "Line" = $MyInvocation.ScriptName
        "ScriptLineNumber" = $MyInvocation.ScriptLineNumber
        "Trace" = $MyInvocation.ScriptStackTrace
    }
    WriteFunction -data $programLogEntry -header @("Timestamp", "Instance", "Table", "LogMessage", "ScriptName", "Line", "ScriptLineNumber", "Trace") -file ".\log\program.log"
}
catch {
    $errorMessage = $($_.ToString().Trim())
    Write-Error "Error: $errorMessage"

    # Write to Program log
    . .\src\utils\fxWriteLog.ps1
    $errorLogEntry = [PSCustomObject]@{
        "Timestamp" = Get-Date
        "Instance" = $InstanceName
        "Table" = "[$Database].[$Schema].[$TableName]"
        "LogMessage" = "Error: $errorMessage"
        "ScriptName" = "$($_.InvocationInfo.ScriptName)"
        "Line" = "$($_.InvocationInfo.Line.Trim())"
        "ScriptLineNumber" = "$($_.InvocationInfo.ScriptLineNumber)"
        "Trace" = "$($_.ScriptStackTrace.Trim())"
    }
    WriteFunction -data $errorLogEntry -header @("Timestamp", "Instance", "Table", "LogMessage", "ScriptName", "Line", "ScriptLineNumber", "Trace") -file ".\log\program.log"
}



src/utils/fxWriteLog.ps1
Function WriteFunction {
    [CmdletBinding()]
    param (
      [Parameter(Mandatory=$true)][pscustomobject]$data, 
      [Parameter(Mandatory=$true)][array]$header, 
      [Parameter(Mandatory=$true)][string]$file
    )
    begin {
      $retryCount = 0
      $maxRetries = 10
    }
    process {
      while ($retryCount -lt $maxRetries) {
        try {
          if (!(Test-Path $file)) {
            $data | Select-Object $header | ConvertTo-Csv -Delimiter "|" -NoTypeInformation `
              | Out-File $file -Encoding utf8
          } else {
            $data | Select-Object $header | ConvertTo-Csv -Delimiter "|" -NoTypeInformation `
              | Select-Object -Skip 1 | Out-File $file -Encoding utf8 -Append
          }        
          break
        } catch [System.IO.IOException] {
          if ($_.Exception.Message -like "*being used by another process*") {
            $retryCount++
            Write-Host "File is locked, retrying in 1 second..."
            Start-Sleep -Seconds 1
          } else {
            Write-Error "Unexpected error: $_"
            break
          }
        }
      }
    }
    end {
      if ($retryCount -eq $maxRetries) {
        Write-Host "Maximum number of retries reached, giving up."
      }
    }
    <#
        .SYNOPSIS
        This function takes in data, header, and a file path as parameters, and writes the data 
        to the specified file with a pipe delimiter, handling any file locking issues by retrying up to 10 times.
    #>
}

src/utils/fxStageData.ps1
Function StageData {
    [CmdletBinding()]
    param(
        [Parameter(Mandatory=$true)][string]$InstanceName,
        [Parameter(Mandatory=$true)][string]$Database,
        [Parameter(Mandatory=$true)][string]$Schema,
        [Parameter(Mandatory=$true)][string]$rawfilepath,
        [Parameter(Mandatory=$true)][string]$convertedfilepath,
        [Parameter(Mandatory=$false)][string]$SqlDataType = "VARCHAR(MAX)",
        [Parameter(Mandatory=$false)][pscredential]$SqlCred,
        [Parameter(Mandatory=$true)][string]$TableName,
        [Parameter(Mandatory=$false)][Switch]$Append
    )
    begin {
        # Check file existence and/or parse to TSV
        try {
            if(-not (Test-Path $rawfilepath)){
                Write-Error "Could not find: $rawfilepath" -ErrorAction Stop
            } 
            elseif((Test-Path $rawfilepath) -and ($rawfilepath -like '*.csv')){          
                $inputFile = $convertedfilepath -replace '\.csv$', '.tsv'

                # Parse CSV to tab-separated file
                Import-Csv -Path $rawfilepath | ConvertTo-Csv -Delimiter "`t" -NoTypeInformation `
                    | ForEach-Object {$_ -replace '"', ""} | Out-File -FilePath $inputFile -Encoding UTF8
                
                # Remove-Item -Path $rawfilepath

                # Retrieve file header
                $Data = Get-Content -Path $inputFile -TotalCount 1
            }
            elseif((Test-Path $rawfilepath) -and ($rawfilepath -like '*.tsv')){
                $inputFile = $rawfilepath
                $Data = Get-Content -Path $inputFile -TotalCount 1
            }
            else{
                Write-Error "Invalid file $rawfilepath" -ErrorAction Stop
            }
        }
        catch {
            $PSCmdlet.ThrowTerminatingError($PSItem)
        }
    }
    process {
        try {
            # Parse and clean up the headers of the tab-separated file.
            $Header = ($Data, $Data | ConvertFrom-Csv -Delimiter "`t").PSObject.Properties.Name
            Write-Verbose "[Raw File Headers] $Header"
            $CleanHeader = @()
            
            foreach($h in $Header){
                $CleanValue = $h -Replace '[^a-zA-Z0-9_]',''
                $CleanHeader += $CleanValue
            }
            Write-Verbose "[Cleaned File Headers] $CleanHeader"
            
            # Build create table statement for staging table
            $StagingTableName = "temp_{0}" -f $TableName
            if(-not $Append){
                $tempsql = @("IF EXISTS (SELECT 1 FROM sys.tables WHERE name  = '$StagingTableName') DROP TABLE [$Database].[$Schema].[$StagingTableName]; `n")
                $tempsql += ("CREATE TABLE [$Database].[$Schema].[$StagingTableName]($($CleanHeader[0]) $SqlDataType `n")
            } else {
                    $tempsql = @("IF NOT EXISTS (SELECT 1 FROM sys.tables WHERE name  = '$StagingTableName') `n")
                    $tempsql += ("CREATE TABLE [$Database].[$Schema].[$StagingTableName]($($CleanHeader[0]) $SqlDataType `n")
            }
            $CleanHeader[1..$CleanHeader.Length] | ForEach-Object {$tempsql += ",$_ $SqlDataType `n"}
            $tempsql += ");"
            $tempsql = $tempsql -join "`n"
            Write-Verbose "[CREATE TABLE Statement] $tempsql"
        
            # Execute create table statement and bulk load data into staging table
            if($SqlCred){
                Invoke-Sqlcmd -ServerInstance $InstanceName -Database $Database -Query $tempsql -Username $SqlCred.UserName `
                    -Password $SqlCred.GetNetworkCredential().Password -ErrorAction Stop
                
                # Invoking the BCP (Bulk Copy Program) utility
                $cmd = "bcp '[$Database].[$Schema].[$StagingTableName]' in '$inputFile' -S'$InstanceName' -F2 -c -t'\t' -U'$($SqlCred.UserName)' -P'$($SqlCred.GetNetworkCredential().Password)'"
            } else {
                Invoke-Sqlcmd -ServerInstance $InstanceName -Database $Database -Query $tempsql -ErrorAction Stop

                # Invoking the BCP (Bulk Copy Program) utility
                $cmd = "bcp '[$Database].[$Schema].[$StagingTableName]' in '$inputFile' -S'$InstanceName' -F2 -c -t'\t' -T"
            }
            Write-Verbose "[BCP Command] $cmd"
            $cmdout = Invoke-Expression $cmd -ErrorAction Stop
        }
        catch {
            $PSCmdlet.ThrowTerminatingError($PSItem)
        }
    }
    end {
        Write-Host $cmdout -Separator "`n" -ForegroundColor Green
        Write-Host "Source data staging complete >>> initializing data integration...." -ForegroundColor Green
        Write-Verbose "[BCP Results] $cmdout"
        return @{'processed_data_headers'=$CleanHeader}
    }
    <#
        .SYNOPSIS
        This function takes a file in CSV or TSV format, cleans and parses its headers, 
        creates a staging table in a SQL Server database, and bulk loads the data from the file 
        into the staging table using the BCP utility.
    #>
}





src/utils/fxIntegrateStagedData.ps1
Function IntegrateData {
    [CmdletBinding()]
    param(
        [Parameter(Mandatory=$true)][string]$InstanceName,
        [Parameter(Mandatory=$true)][string]$Database,
        [Parameter(Mandatory=$true)][string]$Schema,
        [Parameter(Mandatory=$false)][pscredential]$SqlCred,
        [Parameter(Mandatory=$true)][string]$TableName,
        [Parameter(Mandatory=$true)][array]$DataColumns,
        [Parameter(Mandatory=$false)][string]$PrimaryKey = "id"
    )
    begin {
        # Generate update statement body for stored procedure
        $update = @()
        $DataColumns[0..$DataColumns.Length] | ForEach-Object {$update += ",Target.$_ = Source.$_ `n"}
        $update += ";"
        $update = $update -join "`n"
        $UpdateBody = $($update.substring(1))

        # Generate insert statement body for stored procedure
        $insert_schema = @()
        $DataColumns[0..$DataColumns.Length] | ForEach-Object {$insert_schema += ",$_ `n"}
        $insert_schema = $insert_schema -join "`n"
        $InsertBody = $($insert_schema.substring(1))

        $insert_values = @()
        $DataColumns[0..$DataColumns.Length] | ForEach-Object {$insert_values += ",Source.$_ `n"}
        $insert_values = $insert_values -join "`n"
        $ValuesBody = $($insert_values.substring(1))

        $FileVariables=@(
            "Schema=$Schema",
            "TableName=$TableName",
            "PrimaryKey=$PrimaryKey"
        )

        $InputFile = "src/data/transform-data-template.sql"
        $ModifiedInputFile = "src/data/transform-data-actual.sql"

        # Read the SQL script file
        $Script = Get-Content -Path $InputFile -Raw

        # Replace the placeholders with actual values using the format operator
        $ModifiedScript = $Script -f $UpdateBody, $InsertBody, $ValuesBody

        # Write the modified script to the output file
        $ModifiedScript | Out-File -FilePath $ModifiedInputFile -Encoding UTF8

        $Params=@{
            ServerInstance=$InstanceName
            Database=$Database
            InputFile=$ModifiedInputFile
            Variable=$FileVariables
        }
    }
    process {
        try {
            if($SqlCred){
                $resp = Invoke-Sqlcmd @Params -Username $SqlCred.UserName -Password $SqlCred.GetNetworkCredential().Password -ErrorAction Stop
            } else {
                $resp = Invoke-Sqlcmd @Params -ErrorAction Stop
            }
            # Validate response
            if ($resp.Count -eq 0) {
                Write-Error "[$Database].[$Schema].[$TableName] doesn't exist. Refer to src/assets/table_schemas.sql, create table and try again." -ErrorAction Stop
            } else {
                $result_schema = @("Timestamp", "Instance", "Table", "UpdatedRows", "InsertedRows")

                $result = New-Object PSObject -Property @{'Timestamp'=Get-Date;'Instance'=$InstanceName; `
                    'Table'="[$Database].[$Schema].[$TableName]";'UpdatedRows'=$resp.UpdatedRows[0]; `
                    'InsertedRows'=$resp.InsertedRows[1]
                }
            }
        }
        catch {
            $PSCmdlet.ThrowTerminatingError($PSItem)
        }
    }
    end {
        Remove-Item -Path $ModifiedInputFile
        Write-Host "Data integration complete ...." -ForegroundColor Green
        return @{'log'=$result;'log_header'=$result_schema}
    }
    <#
        .SYNOPSIS    
        This function integrates data from a staging table in a SQL Server database into a specified database table 
        by executing an SQL script containing insert and update statements, and returns a PSObject with 
        the updated and inserted row counts along with metadata.
    #>
}

setup.sh
#!/bin/bash

# Create environment variable file for docker-compose environment params
env_var="SA_PASSWORD=P@s5wOrd
DATABASE=testdb
SCHEMA=dbo
TABLE=mock_data
DB_USER=sa
DATASET=mock_data.csv
SQL_SERVER_IP=0.0.0.0,1433"
echo "$env_var" > .env
echo "Environment variables file created in root directory ...."

# Verify that the data/raw dir isn't empty
if [[ $(find data/raw -mindepth 1 -print -quit) ]]; then
    echo "data/raw directory is not empty."
else
    echo "data/raw directory is empty."
    exit 1
fi

# Build and run the sql-server-db service
docker-compose up -d --remove-orphans sql-server-db

sleep 30

# Print to console sql-server-db-IP-ADDRESS
ip_address=$(docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' dc_sql-server-db)

echo "sql-server-db container ip: $ip_address"

export SQL_SERVER_IP=$ip_address

# Print to console sql-server-db-PORT
# port=$(docker inspect -f '{{range $p, $conf := .NetworkSettings.Ports}}{{(index $conf 0).HostPort}} -> {{$p}}{{end}}' dc_sql-server-db)
port=$(docker inspect -f '{{range $p, $conf := .NetworkSettings.Ports}}{{(index $conf 0).HostPort}}{{end}}' dc_sql-server-db)

echo "sql-server-db container host port: $port"

# Build and run the data-integrator service
docker-compose up -d --remove-orphans data-integrator

sleep 30

# -----------------------------------
# Set some environment variables defined in the .env file
eval $(cat .env | grep SA_PASSWORD)
eval $(cat .env | grep DB_USER)

# Invoke sqlcmd in container db
echo "Invoking the sqlcmd terminal in sql-server-db service ...."
echo "See .db/data-load-query.sql for sql command examples to run in terminal ...."
echo "Press Ctrl + C to terminate the process."
docker exec -it dc_sql-server-db /opt/mssql-tools/bin/sqlcmd -S $ip_address,$port -U $DB_USER -P $SA_PASSWORD

# exit